mlir/lib/Dialect/Presburger/Set.cpp

```diff
@@ -0,0 +1,396 @@
+    // we need the complements (strict negative and strict positive).
+    // TODO reimplement the heuristics
+    const auto &eq = S_i.getEquality(j);
+    recurseWithInequalityFromEquality(eq, true, true);
+    recurseWithInequalityFromEquality(eq, false, true);
+
+    addInequalityFromEquality(eq, false, false);
+    addInequalityFromEquality(eq, true, false);
+    addedIneqs += 2;
+  }
+
+  // offset = 2 * S_i.getNumEqualities();
+  for (size_t j = 0; j < S_i.getNumInequalities(); j++) {
+    /*if (isMarkedRedundant[offset + j])
+      continue;*/
+    const auto &ineq = S_i.getInequality(j);
+
+    size_t snap = simplex.getSnapshot();
+
+    SmallVector<int64_t, 64> complement = complementIneq(ineq);
+
+    B.addInequality(complement);
+    simplex.addInequality(complement);
+    subtractRecursively(B, simplex, S, i + 1, result);
+    B.removeInequality(B.getNumInequalities() - 1);
+    simplex.rollback(snap);
+
+    B.addInequality(ineq);
+    addedIneqs++;
+    simplex.addInequality(ineq);
+  }
+
+  for (size_t i = 0; i < addedIneqs; i++)
+    B.removeInequality(B.getNumInequalities() - 1);
+
+  // TODO benchmark technically we can probably drop this as the caller will
+  // rollback. See if it makes much of a difference. Only the last rollback
+  // would be eliminated by this.
+  simplex.rollback(initialSnap);
+}
+
+// Returns the set difference B - S.
+PresburgerSet PresburgerSet::subtract(FlatAffineConstraints c,
+                                      const PresburgerSet &set) {
+  assert(set.getNumDims() == c.getNumDimIds() &&
+         set.getNumSyms() == c.getNumSymbolIds() &&
+         "Sets to be subtracted have different dimensionality");
+  if (c.isEmptyByGCDTest())
+    return PresburgerSet::makeEmptySet(c.getNumDimIds(), c.getNumSymbolIds());
+
+  if (set.isUniverse())
+    return PresburgerSet::makeEmptySet(set.getNumDims(), set.getNumSyms());
+  if (set.isMarkedEmpty())
+    return PresburgerSet(set.getNumDims(), set.getNumSyms());
+
+  PresburgerSet result(c.getNumDimIds());
+  Simplex simplex(c);
+  subtractRecursively(c, simplex, set, 0, result);
+  return result;
+}
+
+PresburgerSet PresburgerSet::complement(const PresburgerSet &set) {
+  return subtract(FlatAffineConstraints(set.getNumDims(), set.getNumSyms()),
+                  set);
+}
+
+// Subtracts the set S from the current set.
+//
+// We compute (U_i T_i) - (U_i S_i) as U_i (T_i - U_i S_i).
+void PresburgerSet::subtract(const PresburgerSet &set) {
+  assert(set.getNumDims() == nDim && set.getNumSyms() == nSym &&
+         "Sets to be subtracted have different dimensionality");
+  assert(set.getNumSyms() + nSym == 0 &&
+         "operations on sets with symbols are not yet supported");
+  if (markedEmpty)
+    return;
+  if (set.isMarkedEmpty())
+    return;
+  if (set.isUniverse()) {
+    markedEmpty = true;
+    return;
+  }
+  if (isUniverse()) {
+    *this = PresburgerSet::complement(set);
```

Comment:
This is unnecessary, right? The same thing will happen below.

Reply:
I don't think so. If `isUniverse()` is true, then `flatAffineConstraints` might be empty. In this case we will just return `result` which is an empty set.

Reply:
Yes, you're right.